triton-client/src/components/DatasetList/index.tsx

```diff
@@ -1,65 +1,105 @@
+import { useAppDispatch } from "@store/hooks"
 import { DatasetListProps } from "./interfaces"
+import { fetchDatasets, fetchReadsets, fetchRequests } from "@store/thunks"
+import {
+    DatasetColumnSource,
+    useDatasetColumns,
+} from "@components/DatasetColumns"
+import { TritonDataset, TritonRequest } from "@api/api-types"
 
-export default function DatasetList({ runName }: DatasetListProps) {
-    const datasetIDs = useAppSelector(
-        (state) => state.runsState.runsByName[runName]?.datasets,
-    )
-    const datasetsByID = useAppSelector(
-        (state) => state.datasetsState.datasetsById,
-    )
+export default function DatasetList({ externalProjectID }: DatasetListProps) {
+    const dispatch = useAppDispatch()
+    const [isFetching, setIsFetching] = useState(true)
+    const [dataSource, setDataSource] = useState<
+        NonNullable<TableProps<DatasetColumnSource>["dataSource"]>
+    >([])
 
-    const renderDatasets = useMemo(() => {
-        if (datasetIDs === undefined) return [<Spin key={"spin"} />]
-        if (datasetIDs.length === 0)
-            return [
-                <Empty
-                    key={"empty"}
-                    description={
-                        "There are no sample data available for request."
-                    }
-                />,
-            ]
-        if (datasetIDs.length > 0) {
-            console.info(datasetIDs, datasetsByID)
+    useEffect(() => {
+        ;(async () => {
+            const datasets = await dispatch(fetchDatasets(externalProjectID))
+            setDataSource(
+                datasets.map((dataset) => ({
+                    id: dataset.id,
+                    lane: dataset.lane,
+                    external_project_id: dataset.external_project_id,
+                    latest_release_update: new Date(
+                        dataset.latest_release_update,
+                    ),
+                    isFetchingRequest: true,
+                    activeRequest: undefined,
+                    totalSize: 0, // size of 0 indicates that the size is not yet fetched
+                })),
+            )
+            setIsFetching(false)
+            const readsets = await dispatch(
+                fetchReadsets(datasets.map((dataset) => dataset.id)),
+            )
+            const requests = await dispatch(
+                fetchRequests(datasets.map((dataset) => dataset.id)),
+            )
+            const totalSizeByDatasets = readsets.reduce<
+                Record<TritonDataset["id"], number>
+            >((acc, readset) => {
+                acc[readset.dataset] =
+                    (acc[readset.dataset] || 0) + readset.total_size
+                return acc
+            }, {})
+            const activeRequestByDatasets = requests.reduce<
+                Record<TritonDataset["id"], TritonRequest | undefined>
+            >((acc, request) => {
+                acc[request.dataset_id] = request
+                return acc
+            }, {})
```

Comment:
this single useEffect hook function has a lot of logic in it. it would be good to create functions in order to compartmentalize the logic into their own functions (the same approach as with the components but at the function level)